File: rabbitmq/csharp/EmailsService/Producer/EmailProducer.cs
using EmailsService.Config;
using EmailsService.Model;
using RabbitMQ.Client;
using System.Text.Json;

namespace EmailsService.Producer;

public class EmailProducer : IDisposable
{
    private readonly IConnection _connection;
    private readonly IModel _channel;
    private readonly ILogger<EmailProducer> _logger;

    public EmailProducer(ILogger<EmailProducer> logger)
    {
        _logger = logger;
        
        var factory = new ConnectionFactory
        {
            HostName = "localhost",
            Port = 5672,
            UserName = "guest",
            Password = "guest"
        };

        _connection = factory.CreateConnection();
        _channel = _connection.CreateModel();
        
        RabbitMQConfig.ConfigureRabbitMQ(_channel);
    }

    public void SendEmail(EmailMessage message)
    {
        try
        {
            _logger.LogInformation("Sending email to queue: {Message}", JsonSerializer.Serialize(message));

            var body = RabbitMQConfig.SerializeMessage(message);

            _channel.BasicPublish(
                exchange: RabbitMQConfig.EmailExchange,
                routingKey: RabbitMQConfig.EmailRoutingKey,
                basicProperties: null,
                body: body
            );

            _logger.LogInformation("Email sent successfully! ID: {Id}", message.Id);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error sending email: {Message}", ex.Message);
            throw new Exception("Failed to send email", ex);
        }
    }

    public void Dispose()
    {
        _channel?.Close();
        _connection?.Close();
        _channel?.Dispose();
        _connection?.Dispose();
    }
}

File: rabbitmq/java/emails/src/main/java/com/edp/emails/producer/EmailProducer.java
package com.edp.emails.producer;

import com.edp.emails.config.RabbitMQConfig;
import com.edp.emails.dto.EmailMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.stereotype.Service;

@Service
public class EmailProducer {
    
    private static final Logger logger = LoggerFactory.getLogger(EmailProducer.class);

    private final RabbitTemplate rabbitTemplate;

    public EmailProducer(RabbitTemplate rabbitTemplate) {
        this.rabbitTemplate = rabbitTemplate;
    }

    public void sendEmail(EmailMessage message) {
        try {
            logger.info("📤 Sending email to queue: {}", message);

            rabbitTemplate.convertAndSend(
                RabbitMQConfig.EMAIL_EXCHANGE,
                RabbitMQConfig.EMAIL_ROUTING_KEY,
                message
            );
            
            logger.info("✅ Email sent successfully! ID: {}", message.id());
            
        } catch (Exception error) {
            logger.error("❌ Error sending email: {}", error.getMessage());
            throw new RuntimeException("Failed to send email", error);
        }
    }
}